### DISCUSSION TITLE
Upload and Download dumps

### DISCUSSION THREAD

**Discussion by ManyTheFish** [meilisearch/product#148] [Feedback & Feature Proposal]
# Summary

Add a route that allows the download of a dump stored in the server.

# Motivations

https://github.com/meilisearch/MeiliSearch/discussions/1006#discussion-24208

# Explanation

download a dump:
**GET**: `/dumps/:dump_uid`
`200`: stream of the file

upload a dump:
**POST**?: `/dumps/:dump_uid`

# Implementation

**Download**:
- using an `HttpResponse` [streaming](https://docs.rs/actix-web/3.1.0/actix_web/dev/struct.HttpResponseBuilder.html#method.streaming) ?
- use actix file [services](https://github.com/actix/examples/blob/master/static_index/src/main.rs) ?

**Upload**:
- https://turreta.com/2019/09/26/rust-upload-files-in-actix-web/ ?
- https://docs.rs/actix-multipart/0.3.0/actix_multipart/ ?

**Comment by ppamorim** [comment]
@LegendreM I wrote down some of my thoughts about this feature.

 - I agree with the routes `GET - /dumps/:dump_uid` and `POST - /dumps/:dump_uid`, but I think the route `POST - /dumps` should be renamed to `POST - /dumps/create` to prevent any confusion with `GET - /dumps/:dump_uid`.

 - I believe that by using `HttpResponse` for both download and upload functions are essential to keep the API stable, by dividing the download function to services and `HttpResponse` to the upload can be confusing for the user. If we use services the user will need to manually find the dump in the services folder instead of simply requesting the dump file. Also, this limits us for new implementation and additional features related to the download of dumps.

Some information:

 - I already implemented the download function with the help from @MarinPostma.
 - I started the development of the uploader but I could not find a relatively easy way to make everything non-blocking by using `actix` + `tokio` and supporting multipart upload at the same time.
